cpp/tests/utilities/scalar_utilities.cu

```diff
@@ -34,34 +35,42 @@ struct compare_scalar_functor
 {
     template<typename T>
     typename std::enable_if_t<not std::is_floating_point<T>::value, void>
-    operator()(cudf::scalar const& lhs, cudf::scalar const& rhs, double tolerance)
+    operator()(cudf::scalar const& lhs, cudf::scalar const& rhs)
     {
         auto lhs_t = static_cast<scalar_type_t<T> const&>(lhs);
         auto rhs_t = static_cast<scalar_type_t<T> const&>(rhs);
         EXPECT_EQ(lhs_t.value(), rhs_t.value());
     }
 
     template<typename T>
-    typename std::enable_if_t<std::is_floating_point<T>::value, void>
-    operator()(cudf::scalar const& lhs, cudf::scalar const& rhs, double tolerance)
+    std::enable_if_t<std::is_same<T, float>::value>
+    operator()(cudf::scalar const& lhs, cudf::scalar const& rhs)
     {
         auto lhs_t = static_cast<scalar_type_t<T> const&>(lhs);
         auto rhs_t = static_cast<scalar_type_t<T> const&>(rhs);
-        EXPECT_NEAR(lhs_t.value(), rhs_t.value(), tolerance);
+        EXPECT_FLOAT_EQ(lhs_t.value(), rhs_t.value());
+    }
+
+    template<typename T>
+    std::enable_if_t<std::is_same<T, double>::value>
```

Comment:
Just for future reference. We make use of `enable_if` a lot in `libcudf` for type dispatched things, but sometimes just plain old specialization can be cleaner. So this could just be:
```c++
template <>
void operator()<double>(...)
```

As a guidance, `enable_if` is useful when you want to enable a particular code path for a variety of types that satisfy some condition. Specialization is useful when you want to select a code path for a _specific_ type.

Reply:
Makes sense, but I think in this case I'd need to forward declare the operator definition and specialize outside of the struct definition. 

```
struct compare_scalar_functor
{
    template<typename T> void operator()(cudf::scalar const& lhs, cudf::scalar const& rhs);
};

template<typename T>
void compare_scalar_functor::operator()(cudf::scalar const& lhs, cudf::scalar const& rhs)
{
    auto lhs_t = static_cast<scalar_type_t<T> const&>(lhs);
    auto rhs_t = static_cast<scalar_type_t<T> const&>(rhs);
    EXPECT_EQ(lhs_t.value(), rhs_t.value());
}

template<>
void compare_scalar_functor::operator()<float>(cudf::scalar const& lhs, cudf::scalar const& rhs)
{
    auto lhs_t = static_cast<scalar_type_t<float> const&>(lhs);
    auto rhs_t = static_cast<scalar_type_t<float> const&>(rhs);
    EXPECT_FLOAT_EQ(lhs_t.value(), rhs_t.value());
}

template<>
void compare_scalar_functor::operator()<double>(cudf::scalar const& lhs, cudf::scalar const& rhs)
{
    auto lhs_t = static_cast<scalar_type_t<double> const&>(lhs);
    auto rhs_t = static_cast<scalar_type_t<double> const&>(rhs);
    EXPECT_DOUBLE_EQ(lhs_t.value(), rhs_t.value());
}
```

If I use three separate templates and filter by trait, I can keep them inside the struct definition (since that doesn't require specialization).

Reply:
As of C++14, specializations can appear within class scope (really, [any scope](https://wg21.cmeerw.net/cwg/issue727)). However, even the latest `gcc` is non-conformant with this rule: see [clang](https://wandbox.org/permlink/8VPnbZ5EivGAVAVC) vs. [gcc](https://wandbox.org/permlink/8VPnbZ5EivGAVAVC). 


Given we're stuck with `gcc`, this doesn't really help you. However, your example is a little more complex than it needs to be. There's no need to forward declare. Only the specializations need to be outside the class (in namespace scope). 
```c++
struct compare_scalar_functor
{
    template<typename T> void operator()(cudf::scalar const& lhs, cudf::scalar const& rhs){
       auto lhs_t = static_cast<scalar_type_t<T> const&>(lhs);
       auto rhs_t = static_cast<scalar_type_t<T> const&>(rhs);
       EXPECT_EQ(lhs_t.value(), rhs_t.value());
};
template<>
void compare_scalar_functor::operator()<float>(cudf::scalar const& lhs, cudf::scalar const& rhs)
{
    auto lhs_t = static_cast<scalar_type_t<float> const&>(lhs);
    auto rhs_t = static_cast<scalar_type_t<float> const&>(rhs);
    EXPECT_FLOAT_EQ(lhs_t.value(), rhs_t.value());
}

template<>
void compare_scalar_functor::operator()<double>(cudf::scalar const& lhs, cudf::scalar const& rhs)
{
    auto lhs_t = static_cast<scalar_type_t<double> const&>(lhs);
    auto rhs_t = static_cast<scalar_type_t<double> const&>(rhs);
    EXPECT_DOUBLE_EQ(lhs_t.value(), rhs_t.value());
}